Can not resolve member user name in the audit log

**Expected behavior and actual behavior:**
1. Username should be resolved in the audit log


**Steps to reproduce the problem:**
1. install harbor with db_auth.
2. login with admin user
3. create user mike01
4. Add user mike01 to library project as admin role.
5. Check the audit log, found the log item

<img width="1914" height="229" alt="Image" src="https://github.com/user-attachments/assets/37684f5f-eef0-480e-823f-3aa619c20ae5" />

the core.log 
```
Jul 14 09:53:04 zdj-dev core[3334048]: 2026-07-14T09:53:04Z [ERROR] [/pkg/auditext/event/member/member.go:195]: failed to resolve project 1: sql: transaction has already been committed or rolled back
Jul 14 09:53:04 zdj-dev core[3334048]: 2026-07-14T09:53:04Z [ERROR] [/pkg/auditext/event/member/member.go:200]: failed to resolve project 1: sql: transaction has already been committed or rolled back
```

**Versions:**
Please specify the versions of following systems.

- harbor version: 2.16.0
- docker engine version: [y.y.y]
- docker-compose version: [z.z.z]

**Additional context:**

--- a/src/pkg/auditext/event/member/member.go
+++ b/src/pkg/auditext/event/member/member.go
@@ -217,8 +217,13 @@ func ensureORMContext(ctx context.Context) context.Context {
 	if ctx == nil {
 		return orm.Context()
 	}
-	if _, err := orm.FromContext(ctx); err == nil {
-		return ctx
+	if o, err := orm.FromContext(ctx); err == nil {
+		if _, ok := o.(beegoorm.TxOrmer); !ok {
+			return ctx
+		}
 	}
+	// Member audit events are resolved asynchronously and may run after the
+	// request transaction has already been committed/rolled back. Replace
+	// transaction-bound ORM with a fresh ORM to avoid using a completed tx.
 	return orm.NewContext(ctx, beegoorm.NewOrm())
 }

--- a/src/pkg/auditext/event/member/member.go
+++ b/src/pkg/auditext/event/member/member.go
@@ -213,6 +213,11 @@ func parsePreResolved(info string) (string, string) {
 	return info, ""
 }
 
+// newBeegoOrm builds a fresh, non-transaction ORM. It is a package-level seam
+// so tests can exercise the transaction-replacement branch of ensureORMContext
+// without a registered default database (beegoorm.NewOrm panics otherwise).
+var newBeegoOrm = beegoorm.NewOrm
+
 func ensureORMContext(ctx context.Context) context.Context {
 	if ctx == nil {
 		return orm.Context()
@@ -225,5 +230,5 @@ func ensureORMContext(ctx context.Context) context.Context {
 	// Member audit events are resolved asynchronously and may run after the
 	// request transaction has already been committed/rolled back. Replace
 	// transaction-bound ORM with a fresh ORM to avoid using a completed tx.
-	return orm.NewContext(ctx, beegoorm.NewOrm())
+	return orm.NewContext(ctx, newBeegoOrm())
 }

--- a/src/pkg/auditext/event/member/member_test.go
+++ b/src/pkg/auditext/event/member/member_test.go
@@ -19,6 +19,8 @@ import (
 	"net/http"
 	"testing"
 
+	beegoorm "github.com/beego/beego/v2/client/orm"
+
 	"github.com/goharbor/harbor/src/controller/event/metadata/commonevent"
 	"github.com/goharbor/harbor/src/controller/event/model"
 	"github.com/goharbor/harbor/src/lib/orm"
@@ -393,13 +395,41 @@ func TestAuditLogMemberEventEnabled_EmptyOperation(t *testing.T) {
 }
 
 func TestEnsureORMContext(t *testing.T) {
-	// context with a regular (non-tx) ORM: returned as-is, covering the new
-	// TxOrmer type-assertion branch added to ensureORMContext.
-	regularCtx := orm.NewContext(context.Background(), &ormtesting.FakeOrmer{})
-	gotCtx := ensureORMContext(regularCtx)
-	if gotCtx != regularCtx {
-		t.Error("ensureORMContext with non-tx ORM should return the same context")
-	}
+	// Stub the ORM factory so the transaction-replacement branch does not need a
+	// registered default DB (beegoorm.NewOrm panics otherwise).
+	origNewOrm := newBeegoOrm
+	defer func() { newBeegoOrm = origNewOrm }()
+	fresh := &ormtesting.FakeOrmer{}
+	newBeegoOrm = func() beegoorm.Ormer { return fresh }
+
+	t.Run("non-tx ORM returned as-is", func(t *testing.T) {
+		// A regular (non-tx) ORM must be returned unchanged, covering the
+		// TxOrmer type-assertion branch in ensureORMContext.
+		ctx := orm.NewContext(context.Background(), &ormtesting.FakeOrmer{})
+		if got := ensureORMContext(ctx); got != ctx {
+			t.Error("ensureORMContext with non-tx ORM should return the same context")
+		}
+	})
+
+	t.Run("tx ORM replaced with fresh non-tx ORM", func(t *testing.T) {
+		// A transaction ORM must be swapped for a fresh non-tx ORM, guarding the
+		// regression where member audit events reused a completed transaction.
+		ctx := orm.NewContext(context.Background(), &ormtesting.FakeTxOrmer{})
+		got := ensureORMContext(ctx)
+		if got == ctx {
+			t.Fatal("ensureORMContext with tx ORM should return a new context")
+		}
+		o, err := orm.FromContext(got)
+		if err != nil {
+			t.Fatalf("orm.FromContext returned error: %v", err)
+		}
+		if _, isTx := o.(beegoorm.TxOrmer); isTx {
+			t.Error("ensureORMContext should replace the transaction ORM with a non-tx ORM")
+		}
+		if o != fresh {
+			t.Error("ensureORMContext should install the ORM produced by newBeegoOrm")
+		}
+	})
 }
 
 func TestParsePreResolved(t *testing.T) {

--- a/src/pkg/auditext/event/member/member_test.go
+++ b/src/pkg/auditext/event/member/member_test.go
@@ -21,7 +21,9 @@ import (
 
 	"github.com/goharbor/harbor/src/controller/event/metadata/commonevent"
 	"github.com/goharbor/harbor/src/controller/event/model"
+	"github.com/goharbor/harbor/src/lib/orm"
 	"github.com/goharbor/harbor/src/pkg/notifier/event"
+	ormtesting "github.com/goharbor/harbor/src/testing/lib/orm"
 )
 
 func stubLookups() func() {
@@ -387,6 +389,16 @@ func TestResolver_Resolve_NilInputs(t *testing.T) {
 func TestAuditLogMemberEventEnabled_EmptyOperation(t *testing.T) {
 	if auditLogMemberEventEnabled(context.Background(), "") {
 		t.Fatal("auditLogMemberEventEnabled() = true, want false for empty operation")
+	}
+}
+
+func TestEnsureORMContext(t *testing.T) {
+	// context with a regular (non-tx) ORM: returned as-is, covering the new
+	// TxOrmer type-assertion branch added to ensureORMContext.
+	regularCtx := orm.NewContext(context.Background(), &ormtesting.FakeOrmer{})
+	gotCtx := ensureORMContext(regularCtx)
+	if gotCtx != regularCtx {
+		t.Error("ensureORMContext with non-tx ORM should return the same context")
 	}
 }
 

